src/main/java/com/puppycrawl/tools/checkstyle/utils/AnnotationUtil.java

```diff
@@ -123,20 +123,88 @@ public static boolean containsAnnotation(DetailAST ast, Set<String> annotations)
 
         if (!annotations.isEmpty()) {
             final DetailAST firstMatchingAnnotation = findFirstAnnotation(ast, annotationNode -> {
-                DetailAST identNode = annotationNode.findFirstToken(TokenTypes.IDENT);
-                if (identNode == null) {
-                    identNode = annotationNode.findFirstToken(TokenTypes.DOT)
-                            .findFirstToken(TokenTypes.IDENT);
-                }
-
-                return annotations.contains(identNode.getText());
+                return doesAnnotationMatch(annotationNode, annotations);
             });
             result = firstMatchingAnnotation != null;
         }
 
         return result;
     }
 
+    /**
+     * Checks if annotaton is found in the set of provided annotations.
+     *
+     * @param annotations A collection of annotations to look through.
+     * @param annotationNode Annotation to look for.
+     * @return {@code true} if the annotation is contained in the set.
+     */
+    private static boolean doesAnnotationMatch(DetailAST annotationNode, Set<String> annotations) {
+        final DetailAST identNode = annotationNode.findFirstToken(TokenTypes.IDENT);
+        final String annotationString;
+
+        // If no `IDENT` is found, then we have a `DOT` -> more than 1 qualifier
+        if (identNode == null) {
+            final DetailAST dotNode = annotationNode.findFirstToken(TokenTypes.DOT);
+            annotationString = convertAnnotationNameToString(dotNode);
+        }
+        else {
+            annotationString = identNode.getText();
+        }
+
+        boolean annotationsMatch = false;
+
+        for (String annotation : annotations) {
+            if (annotation.equals(annotationString)) {
+                annotationsMatch = true;
+                break;
+            }
+        }
+
+        return annotationsMatch;
+    }
+
+    /**
+     * Converts (qualified) annotation name to string.
+     *
+     * <p>
+     * For example:
+     * </p>
+     * <pre>
+     * &#64;org.example.MyAnn -> "org.example.MyAnn"
+     * &#64;MyAnn -> "MyAnn"
+     * </pre>
+     *
+     * @param rootDotNode The root DOT node of the (qualified) annotation.
+     * @return Annotation name as string.
+     */
+    private static String convertAnnotationNameToString(DetailAST rootDotNode) {
```

Comment:
Can we use `FullIdent#createFullident` instead of creating new method?

Reply:
Now that I am trying it, `FullIdent` was what I needed. I tried looking for it in other package or import checks but could not find it.

Reply:
It seems that `FullIdent#createFullIdent()` does not pass the following test:
https://github.com/checkstyle/checkstyle/blob/c83780898188a21455c007f491c6d9452d608ea1/src/test/java/com/puppycrawl/tools/checkstyle/utils/AnnotationUtilTest.java#L229-L248
I had issues with this test too.

Isn't this test wrong though? The `modifiersAst` it provides literally translates to `Override.`, not `Override`.

Reply:
All tests that are based on custom created ast are illegal. You can remove it, and make coverage by input files

Reply:
But all the `*UtilTest.java` files are somewhat based on custom created ast. Does that mean they have to be updated?

I will update this test so it checks a real case for now.

Reply:
Try to avoid involving in bigger refactoring.
Extend this test case to serve you need if it reasonable.